src/TimingData.h

```diff
@@ -190,50 +210,72 @@ class TimingData
 		return GetSegmentAtRow( BeatToNoteRow(fBeat), tst );
 	}
 
-	#define DefineSegmentWithName(Seg, SegName, SegType) \
-		const Seg* Get##Seg##AtRow( int iNoteRow ) const \
-		{ \
-			const TimingSegment *t = GetSegmentAtRow( iNoteRow, SegType ); \
-			return To##SegName( t ); \
-		} \
-		Seg* Get##Seg##AtRow( int iNoteRow ) \
-		{ \
-			return const_cast<Seg*> (((const TimingData*)this)->Get##Seg##AtRow(iNoteRow) ); \
-		} \
-		const Seg* Get##Seg##AtBeat( float fBeat ) const \
-		{ \
-			return Get##Seg##AtRow( BeatToNoteRow(fBeat) ); \
-		} \
-		Seg* Get##Seg##AtBeat( float fBeat ) \
-		{ \
-			return const_cast<Seg*> (((const TimingData*)this)->Get##Seg##AtBeat(fBeat) ); \
-		} \
-		void AddSegment( const Seg &seg ) \
-		{ \
-			AddSegment( &seg ); \
-		}
-
-	// "XXX: this comment (and quote mark) exists so nano won't
-	// display the rest of this file as one giant string
-
-	// (TimeSignature,TIME_SIG) -> (TimeSignatureSegment,SEGMENT_TIME_SIG)
-	#define DefineSegment(Seg, SegType ) \
-		DefineSegmentWithName( Seg##Segment, Seg, SEGMENT_##SegType )
-
-	DefineSegment( BPM, BPM );
-	DefineSegment( Stop, STOP );
-	DefineSegment( Delay, DELAY );
-	DefineSegment( Warp, WARP );
-	DefineSegment( Label, LABEL );
-	DefineSegment( Tickcount, TICKCOUNT );
-	DefineSegment( Combo, COMBO );
-	DefineSegment( Speed, SPEED );
-	DefineSegment( Scroll, SCROLL );
-	DefineSegment( Fake, FAKE );
-	DefineSegment( TimeSignature, TIME_SIG );
-
-	#undef DefineSegmentWithName
-	#undef DefineSegment
+	// Timing segment management and conversion
+	const BPMSegment* GetBPMSegmentAtRow(int iNoteRow) const { const TimingSegment* t = GetSegmentAtRow(iNoteRow, SEGMENT_BPM); return ToBPM(t); }
```

Comment:
This is still a crazy amount of code duplication:

Try and do something like the following (only doing the non-const versions for simplicity).


```cpp
template<typename T>
inline T* ToSegment(TimingSegment* t, TimingSegmentType expectedType)
{
  ASSERT(t->GetType() == expectedType);
	return static_cast<T*>(t);
}

template <typename T>
T* GetSegmentAtRowTyped(int iNoteRow, TimingSegmentType type)
{
  TimingSegment* t = GetSegmentAtRow(iNoteRow, type);  // Get the base TimingSegment*
  return ToSegment<T>(t, type);                        // Cast it to the correct subclass
}

// Then we can call it like:
BPMSegment* bpmSegment = GetSegmentAtRowTyped<BPMSegment>(iNoteRow, SEGMENT_BPM);
ScrollSegment* scrollSegment = GetSegmentAtRowTyped<ScrollSegment>(iNoteRow, SEGMENT_SCROLL);
StopSegment* stopSegment = GetSegmentAtRowTyped<StopSegment>(iNoteRow, SEGMENT_STOP);
```

Reply:
at a quick glance without having looked at the code very much, the `expectedType` and `T` look redundant. is the `ASSERT` helpful at all?

Reply:
Honestly I feel like the ASSERT is likely redundant, if it wasn't the game would've likely crashed anyways. My guess is it's to guarantee that new timingsegments have a corresponding enum and the GetType function implemented, but I think it's fine.

I think we can get away by just doing:

```cpp
BPMSegment* bpmSegment = static_cast<BPMSegment*>(GetSegmentAtRow(iNoteRow, SEGMENT_BPM));
ScrollSegment* scrollSegment = static_cast<ScrollSegment*>(GetSegmentAtRow(iNoteRow, SEGMENT_SCROLL));
StopSegment* stopSegment = static_cast<StopSegment*>(GetSegmentAtRow(iNoteRow, SEGMENT_STOP));
```